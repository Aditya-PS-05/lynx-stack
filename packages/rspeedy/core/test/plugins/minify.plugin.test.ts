// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
/* eslint-disable vitest/expect-expect */

import {
  CssMinimizerWebpackPlugin,
  pluginCssMinimizer,
} from '@rsbuild/plugin-css-minimizer'
import { beforeEach, describe, expect, test, vi } from 'vitest'

import type { Output } from '../../src/index.js'
import { createStubRspeedy } from '../createStubRspeedy.js'

describe('Plugins - Minify', () => {
  beforeEach(() => {
    vi.restoreAllMocks()
    vi.stubEnv('NODE_ENV', 'production')
    return () => vi.unstubAllEnvs()
  })

  test('defaults', async () => {
    const rsbuild = await createStubRspeedy({})

    const config = await rsbuild.unwrapConfig()

    expect(config.optimization?.minimizer).toMatchSnapshot()
  })

  test('minify: false', async () => {
    const rsbuild = await createStubRspeedy({
      output: { minify: false },
    })

    const config = await rsbuild.unwrapConfig()

    expect(config.optimization?.minimizer).toMatchInlineSnapshot(`undefined`)
  })

  test('minify.compress', async () => {
    const options: Output['minify'][] = [
      { jsOptions: { minimizerOptions: { compress: true } } },
      { jsOptions: { minimizerOptions: { compress: false } } },
      {
        jsOptions: {
          minimizerOptions: {
            compress: {
              expression: false,
              inline: 1,
              pure_funcs: [],
              negate_iife: true,
              dead_code: true,
            },
          },
        },
      },
    ]
    await options.reduce(
      (prev, curr) => prev.then(() => expectMinify(curr)),
      Promise.resolve(),
    )
  })

  test('minify.mangle', async () => {
    const options: Output['minify'][] = [
      { jsOptions: { minimizerOptions: { mangle: true } } },
      { jsOptions: { minimizerOptions: { mangle: false } } },
      {
        jsOptions: {
          minimizerOptions: {
            mangle: {
              props: {},
              toplevel: true,
              reserved: ['foo', 'lynx'],
            },
          },
        },
      },
    ]
    await options.reduce(
      (prev, curr) => prev.then(() => expectMinify(curr)),
      Promise.resolve(),
    )
  })

  test('minify.format', async () => {
    const options: Output['minify'][] = [
      { jsOptions: { minimizerOptions: { format: {} } } },
      {
        jsOptions: {
          minimizerOptions: {
            format: {
              ascii_only: true,
              asciiOnly: false,
              preserve_annotations: true,
            },
          },
        },
      },
    ]

    await options.reduce(
      (prev, curr) => prev.then(() => expectMinify(curr)),
      Promise.resolve(),
    )
  })

  test('minify.extractComments', async () => {
    const options: Output['minify'][] = [
      { jsOptions: { extractComments: {} } },
      { jsOptions: { extractComments: false } },
      { jsOptions: { extractComments: /@license/ } },
      {
        jsOptions: {
          extractComments: {
            condition: /^@keep/,
            banner: 'foo',
          },
        },
      },
      {
        jsOptions: {
          extractComments: {
            condition: false,
            banner: false,
          },
        },
      },
    ]

    await options.reduce(
      (prev, curr) => prev.then(() => expectMinify(curr)),
      Promise.resolve(),
    )
  })

  describe('CSS', () => {
    test('css: false', async () => {
      const rsbuild = await createStubRspeedy({
        output: { minify: { css: false } },
      })

      const config = await rsbuild.unwrapConfig()

      expect(config.optimization?.minimizer).toHaveLength(1)
    })

    test('custom css minimizer', async () => {
      const rsbuild = await createStubRspeedy({
        plugins: [
          pluginCssMinimizer({
            pluginOptions: {
              minify: CssMinimizerWebpackPlugin.cleanCssMinify,
              minimizerOptions: {
                // @ts-expect-error mock
                level: {
                  1: {
                    roundingPrecision: 'all=3,px=5',
                  },
                },
              },
            },
          }),
        ],
      })

      const config = await rsbuild.unwrapConfig()

      expect(config.optimization?.minimizer).toMatchSnapshot()
    })
  })
})

async function expectMinify(minify: Output['minify']): Promise<void> {
  const rsbuild = await createStubRspeedy({ output: { minify } })
  const config = await rsbuild.unwrapConfig()
  expect(config.optimization?.minimizer).toMatchSnapshot()
}
