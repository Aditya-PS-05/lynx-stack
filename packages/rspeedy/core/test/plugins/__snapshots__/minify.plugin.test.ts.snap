// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Plugins - Minify > CSS > custom css minimizer 1`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "level": {
            "1": {
              "roundingPrecision": "all=3,px=5",
            },
          },
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > defaults 1`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.compress 1`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": true,
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.compress 2`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": false,
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.compress 3`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "dead_code": true,
            "ecma": 2015,
            "expression": false,
            "inline": 1,
            "join_vars": false,
            "negate_iife": true,
            "pure_funcs": [],
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.extractComments 1`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": {},
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.extractComments 2`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.extractComments 3`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": /@license/,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.extractComments 4`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": {
          "banner": "foo",
          "condition": /\\^@keep/,
        },
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.extractComments 5`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": {
          "banner": false,
          "condition": false,
        },
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.format 1`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.format 2`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "ascii_only": true,
            "comments": false,
            "keep_quoted_props": true,
            "preserve_annotations": true,
          },
          "mangle": {
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.mangle 1`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": true,
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.mangle 2`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": false,
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;

exports[`Plugins - Minify > minify.mangle 3`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_args": [
      {
        "extractComments": false,
        "minimizerOptions": {
          "compress": {
            "comparisons": false,
            "ecma": 2015,
            "inline": 2,
            "join_vars": false,
            "negate_iife": false,
            "side_effects": false,
            "toplevel": true,
          },
          "format": {
            "asciiOnly": false,
            "comments": false,
            "keep_quoted_props": true,
          },
          "mangle": {
            "props": {},
            "reserved": [
              "foo",
              "lynx",
            ],
            "toplevel": true,
          },
        },
      },
    ],
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  CssMinimizerPlugin {
    "options": {
      "exclude": undefined,
      "include": undefined,
      "minimizer": {
        "implementation": [Function],
        "options": {
          "preset": [
            "default",
            {
              "mergeLonghand": false,
            },
          ],
        },
      },
      "parallel": true,
      "test": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
      "warningsFilter": [Function],
    },
  },
]
`;
